### ISSUE TITLE
Persist pdf data in Milvus for later use

### ISSUE THREAD

**Issue by skprasadu** [hwchase17/langchain#3471]
Using Langchain, I used Milvus vector db to ingest all my document as per @ https://python.langchain.com/en/latest/modules/indexes/vectorstores/examples/milvus.html. Now later I want to get a handle to vector_db and start to query Milvus. How do I achieve this? In the example in the link, it is querying the vector_db immediately. Imagine, once I call from_documents, I set the vector_db=None and later I need to load the collection back and query. How do I do that for Milvus?

**Comment by Charlygnusa** [comment]
After inserting data into Milvus, you can use this code to connect with Milvus, where "collection_name" is the collection you have already inserted into Milvus, and "index_params" is the field name you want to search:

 vector_db = Milvus(
            embeddings,
            connection_args={"host": "127.0.0.1", "port": "19530"},
            collection_name="sample",
            index_params="text"
        )

**Comment by rodgermoore** [comment]
> After inserting data into Milvus, you can use this code to connect with Milvus, where "collection_name" is the collection you have already inserted into Milvus, and "index_params" is the field name you want to search:
> 
> ```
> vector_db = Milvus(
>            embeddings,
>            connection_args={"host": "127.0.0.1", "port": "19530"},
>            collection_name="sample",
>            index_params="text"
>        )
> ```

This doesn't work because of [this in the Langchain Milvus implementation](https://github.com/hwchase17/langchain/blob/master/langchain/vectorstores/milvus.py):

````
    def _init(
        self, embeddings: Optional[list] = None, metadatas: Optional[list[dict]] = None
    ) -> None:
        if embeddings is not None:
            self._create_collection(embeddings, metadatas)
        self._extract_fields()
        self._create_index()
        self._create_search_params()
        self._load()
````

**Comment by denissabramovs** [comment]
Yeah... i'm also searching how to make it. All examples in langchain docs show how to perform short term use, but none for long term use cases.

**Comment by martinholecekmax** [comment]
> > After inserting data into Milvus, you can use this code to connect with Milvus, where "collection_name" is the collection you have already inserted into Milvus, and "index_params" is the field name you want to search:
> > ```
> > vector_db = Milvus(
> >            embeddings,
> >            connection_args={"host": "127.0.0.1", "port": "19530"},
> >            collection_name="sample",
> >            index_params="text"
> >        )
> > ```
> 
> This doesn't work because of [this in the Langchain Milvus implementation](https://github.com/hwchase17/langchain/blob/master/langchain/vectorstores/milvus.py):
> 
> ```
>     def _init(
>         self, embeddings: Optional[list] = None, metadatas: Optional[list[dict]] = None
>     ) -> None:
>         if embeddings is not None:
>             self._create_collection(embeddings, metadatas)
>         self._extract_fields()
>         self._create_index()
>         self._create_search_params()
>         self._load()
> ```
> 
> The Milvus wrapper will try to create a new index and load it while not needed.

Hello Guys,

I went through the code, and I'm pretty sure that you don't have to pass the index_params argument a second time once you already created the collection. It will automatically use the current collection index_params. If you pass the index_params, it will use these index parameters, but you will also have to change search_params to match the index_params.

Here is a short example using hugging face embeddings.

## Create a new collection from texts using hugging face embeddings.
```python
from langchain.vectorstores import Milvus
from langchain.embeddings import HuggingFaceEmbeddings

import os

os.environ["HUGGINGFACEHUB_API_TOKEN"] = "your_huggingface_api_key"

HUGGING_FACE_MODEL_ID = "bert-base-uncased"
MILVUS_COLLECTION_NAME = "test_collection"

embeddings = HuggingFaceEmbeddings(model_name=HUGGING_FACE_MODEL_ID)

texts = ["foo", "bar", "baz"]

# Create a new collection from texts
vector_db: Milvus = Milvus.from_texts(
    texts,
    embedding=embeddings,
    collection_name=MILVUS_COLLECTION_NAME,
    connection_args={"host": "127.0.0.1", "port": "19530"},
)

# Search for similar documents
docs = vector_db.similarity_search(query="foo", k=1)

print("===============================================================")
print("docs: ", docs)
print("===============================================================")
```

## Load existing collection and perform a search
```python
from langchain.vectorstores import Milvus
from langchain.embeddings import HuggingFaceEmbeddings

import os

os.environ["HUGGINGFACEHUB_API_TOKEN"] = "your_huggingface_api_key"

HUGGING_FACE_MODEL_ID = "bert-base-uncased"
MILVUS_COLLECTION_NAME = "test_collection"
embeddings = HuggingFaceEmbeddings(model_name=HUGGING_FACE_MODEL_ID)

# Get existing collection from Milvus
vector_db: Milvus = Milvus(
    embeddings,
    collection_name=MILVUS_COLLECTION_NAME,
    connection_args={"host": "127.0.0.1", "port": "19530"},
)

# Search for similar documents
docs = vector_db.similarity_search(query="foo", k=1)

print("===============================================================")
print("docs: ", docs)
print("===============================================================")
```

Here is the code if you wish to update the index_params after the collection is already created, but I wouldn't recommend doing this. A better solution would be to create a new collection with the index parameters that you want and generate new embeddings for all documents.

```python
# Get existing collection from Milvus with changing index parameters
vector_db: Milvus = Milvus(
    embeddings,
    collection_name=MILVUS_COLLECTION_NAME,
    index_params={
        "metric_type": "IP",
        "index_type": "FLAT",
    },
    search_params={"metric_type": "IP"},
    connection_args={"host": "127.0.0.1", "port": "19530"},
)
```

**Comment by fabmeyer** [comment]
I want to connect to a given Milvus DB without having to add `'texts'` and `'embedding'` again. This makes no sense.

`
from_texts() missing 2 required positional arguments: 'texts' and 'embedding'
`

Any news on this?

**Comment by martinholecekmax** [comment]
Hi @fabmeyer,

You can use constructor if you already have collection.

```py
# Get existing collection from Milvus with changing index parameters
vector_db: Milvus = Milvus(
    embedding_function=embeddings,
    collection_name=MILVUS_COLLECTION_NAME,
    index_params={
        "metric_type": "IP",
        "index_type": "FLAT",
    },
    search_params={"metric_type": "IP"},
    connection_args={"host": "127.0.0.1", "port": "19530"},
)
```

**Comment by fabmeyer** [comment]
@martinholecekmax 
Thanks for the reply. Isn't it possible to connect to an existing Milvus DB without having to provide an embedding function?

**Comment by martinholecekmax** [comment]
Thank you for your question. In the context of working with Milvus, it's important to note that embeddings play a crucial role. The embeddings are used to convert your data into a format that Milvus can understand and work with, which is crucial for conducting vector similarity searches.

While you can certainly connect to an existing Milvus DB without providing an embedding function, you wouldn't be able to perform any meaningful operations or run queries on your data unless it has been embedded. The vector embeddings are essentially the 'language' that Milvus uses to process and compare data, and without them, the functionality of the database would be significantly limited.

Therefore, the provision of an embedding function is not optional in the code implementation. Without it, the core functionality that Milvus provides, such as similarity search and nearest neighbor search, would not be possible. It's like trying to read a book without knowing the language it's written in; you may have access to the book, but without understanding the language, the content will be incomprehensible. The same goes for Milvus without embeddings.

**Comment by martinholecekmax** [comment]
Continuing from my previous point, @fabmeyer, while you can technically use 'fake' embeddings from this [link](https://python.langchain.com/docs/modules/data_connection/text_embedding/integrations/fake), they lack the semantic value necessary for effective operations like similarity searches in Milvus. Even though connection to the database is possible, the lack of genuine embeddings reflecting your data will render your query results meaningless. Therefore, the use of authentic embeddings remains indispensable. I hope this explanation makes the concept clearer.
